### PR TITLE
API: Ensure set_index always sorts the index

### LINKED ISSUE #2288: Keyerror when slicing dating
This looks similar to #2211 but I'm not sure. I've attached a zip with sample data and code that reproduces it. If you uncomment line 25 it works for some reason.
```
import dask.bag
import pandas as pd
import re
from datetime import datetime
schema_dict = {
    'timestamp': 'datetime64[ns]',
}


time_regex = r'\[(?P<time>[^]]+)\]'
time_regex = re.compile(time_regex)


def get_log_dict(line):
    match = time_regex.match(line)
    dt = pd.datetime.strptime(match.groupdict()['time'], '%d/%b/%Y:%H:%M:%S +0000')
    return {'timestamp': dt}


files = ['2012-09-25.log', '2012-09-26.log', '2012-09-27.log']
b = dask.bag.read_text(files, blocksize=5000000).map(get_log_dict).to_dataframe(schema_dict)
b = b[~b.timestamp.isnull()]
b = b.set_index('timestamp')
b = b[sorted(b.columns)]
# b = b.repartition(freq='15m')
start = datetime(2012, 9, 26)
end = datetime(2012, 9, 27)
b = b.loc[start:end]
b.compute()
```

[Archive.zip](https://github.com/dask/dask/files/971843/Archive.zip)

### PR DESCRIPTION
**Pull request by TomAugspurger**
There are edge cases (like the test case here) where we find that things are sorted *between* partitions, but don't check that they are sorted *within* partitions.

This fixes the problem by just calling `sort_index` on each partition. There's might be a more elegant / efficient solution where we can avoid calling `sort_index` in some cases, but I haven't been able to find it yet.

Closes #2288

### CODE REVIEW

File: dask/dataframe/tests/test_dataframe.py [hunk 1]
```diff
@@ -2749,7 +2749,6 @@ def test_slice_on_filtered_boundary(drop):
     result = dd.concat([ddf, ddf.rename(columns={"B": "C"})], axis=1)
     expected = pd.concat([pdf, pdf.rename(columns={"B": "C"})], axis=1)
     assert_eq(result, expected)
-    assert not result.compute().index.is_monotonic  # didn't accidentally sort
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_dataframe.py]: This is an unrelated test that I had to fix. It's now out of date, since `set_index` will sort.

**Comment by mrocklin** [comment]
+1 from me
